src/main/java/org/jabref/logic/importer/fetcher/MergingIdBasedFetcher.java

```diff
@@ -0,0 +1,132 @@
+
+package org.jabref.logic.importer.fetcher;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import javax.swing.undo.UndoManager;
+
+import org.jabref.gui.DialogService;
+import org.jabref.gui.preferences.GuiPreferences;
+import org.jabref.gui.undo.NamedCompound;
+import org.jabref.logic.importer.IdBasedFetcher;
+import org.jabref.logic.importer.ImportCleanup;
+import org.jabref.logic.importer.MergeEntriesHelper;
+import org.jabref.logic.importer.fetcher.isbntobibtex.IsbnFetcher;
+import org.jabref.logic.l10n.Localization;
+import org.jabref.logic.util.BackgroundTask;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.Field;
+import org.jabref.model.entry.field.StandardField;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MergingIdBasedFetcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MergingIdBasedFetcher.class);
+    private final GuiPreferences preferences;
+    private final UndoManager undoManager;
+    private final BibDatabaseContext bibDatabaseContext;
+    private final DialogService dialogService;
+
+    public MergingIdBasedFetcher(GuiPreferences preferences, UndoManager undoManager, BibDatabaseContext bibDatabaseContext, DialogService dialogService) {
+        this.preferences = preferences;
+        this.undoManager = undoManager;
+        this.bibDatabaseContext = bibDatabaseContext;
+        this.dialogService = dialogService;
+    }
+
+    public BackgroundTask<List<String>> fetchAndMergeBatch(List<BibEntry> entries) {
+        if (entries.isEmpty()) {
+            return BackgroundTask.wrap(() -> List.of());
+        }
+
+        BackgroundTask<List<String>> backgroundTask = new BackgroundTask<>() {
```

Comment:
The fetcher itself should not have UI dependencies and implement the background task, that is a UI related component. This has to be done on the caller SIde which I presume is somewhere in the UI

Reply:
Thank you for the feedback!
Will refactor to move UI dependencies to the caller side and update the PR in these two days.

Reply:
Yeah, the wrapping can be done in the caller.